### ISSUE TITLE
Better handling of unrecognized joysticks

### ISSUE THREAD

**Issue by pqftgs** [UPBGE/blender#389]
Rather than throwing an error, could they instead be marked as unrecognized while still letting you read the raw inputs?

Reasons:
1.  There will inevitably be more unsupported controllers as the game ages.  Custom calibration pretty much future-proofs the game.
2.  Lets you support those weird controllers that don't really map to xbox
3.  _Nicer_ than shipping a 3rd-party mapping utility

**Comment by youle31** [comment]
hey! Actually we were talking about that very recently. It just needs to edit this: E:\Compilation\Blender\upbgeSources\source\gameengine\Device\DEV_JoystickMappingdb.h as far I understood. So we'd need to find a protocol to:

- Invite the users to find their joysticks infos on all platforms.
- Make builds so they can test and confirm it works at least on their platform.

**Comment by pqftgs** [comment]
I guess what I'm asking is for better handling of un-mapped controllers (changed title to match).  Even with a 100% complete database there's always gonna be more over time.  And old games probably won't be getting those updates.

Urho3D's joystick class has IsController() which returns true for standard mapped controllers, false for everything else, letting you read inputs even if it's not mapped.  Something like that would be perfect.

That aside, inviting users to improve the joystick database is a good idea.  I might have a couple that need to be mapped...

**Comment by lordloki** [comment]
Hi,
the system is though this way:
- If you have a game in Steam it uses the joystick mapping from Steam Big Mode
- if you don't have the game in Steam it use by defect the joystick mapping passed through environment variable SDL_GAMECONTROLLERCONFIG. (i.e export SDL_GAMECONTROLLERCONFIG="[the string you received from controllermap]"
- if you don't use the SDL_GAMECONTROLLERCONFIG it uses the internal controller database (updated very often from https://github.com/gabomdq/SDL_GameControllerDB)
- if your controller is not recognized by any of those previous method then through an error asking for mapping pointing where you can obtain the apps for this task

If we support read input for not recognized controller we would need duplicate the current system with former joystick API and after mix it with the current one. I though about it at the beginning but the code was a spaghetti code. Anyway I think it will be a problem solved by the time (the gamecontroller database increase its size quickly)

**Comment by youle31** [comment]
wow I didn't know that it was so complex!

**Comment by BluePrintRandom** [comment]
Thanks LordLoki! epic.

**Comment by lordloki** [comment]
Yes ;).

Regarding old games not receiving controller database updates, they will have always the possibility to pass the mapping through the SDL game controller environment variable.
I will update the joystick mediawiki documentation adding previous information.

Anyway, @ pqftgs if you pass me the mapping of your joysticks I will make a pull request to include them in general joystick database

**Comment by pqftgs** [comment]
Gotcha, so basically a non-steam game will need an external configuration tool.  Not perfect but that's good enough for me.  Will close this request.

@lordloki I'll send you some later.  Seems my trigger output is inverted on one of em... I'll open a new issue if I can't get that figured out.

**Comment by lordloki** [comment]
I will add in the next days another way to check and load an external database file. Basically, if a "gamecontrollerdb.txt" file exists in the same directory as blenderplayer one it will load it . 
This way, the user or the game developer can have updated the supported joystick's database without replace the blenderplayer executable by the last version.

**Comment by lordloki** [comment]
Hi @totterfree 

Some time ago we decided to go with this new system that provided SDL2 (and that is becoming a standard as it has direct support from Steam).

The problem is basically that both systems run in parallel but cannot take advantage of each other's characteristics:

- Old system:
 Detects almost every joystick
 It doesn't have hotpluggin
 It's not vibrating.
 Not all buttons are configured equally (in one gamepad the 1 button may be the A button and in another the X button)
 
- New system:
 Detects only the joysticks that are in the database, the others are ignored. (Except if you have the game in Steam as it defaults to your joystick settings on the platform)
 It's got hotpluggin'.
 It's got a vibration.
 Configures all buttons equally (To standard XBox360)
 
Mixing both systems is possible but:
- It would be much more complicated to maintain
- There would be a very large mix of bug reports (i.e., joystick detecting hotpluggin other not, same for vibration, etc)

Therefore, although it will exist corner cases (as yours) where the support is not going to be good we prefer to follow with the new system as it simplyfies joystick handle.

Possible solution:
- Use Pygame library to handle joystick 

Addittionaly, Could you paste here the new controllers that you have added to gamecontrollerdb.txt to can add them to general gamecontrollerdb.txt? 

Thanks in advance

**Comment by organicpencil** [comment]
Damn, forgot all about this. Anyhow - some games have a setting that disables xinput and lets players configure their own unsupported controllers. Pretty nice since you can't rely on devs to update the controller db forever. It seems especially important for BGE since (AFAIK) it gets used for a lot of weird homebrew stuff.

But:
-Coding support for both = unpaid effort
-There's already multiple workarounds
-Obscure workarounds are part of the BGE experience
